### DISCUSSION TITLE
Seperate `Provider` from context

### DISCUSSION THREAD

**Discussion by zhaoyao91** [solidjs/solid#490] [Ideas]
Days ago I made a function `provideContext`, which connects solid context to general sync functions, not only components:

```
import { createMemo, getOwner } from "solid-js";

interface Context<T> {
  id: symbol;
  Provider: (props: { value: T; children: any }) => any;
  defaultValue?: T;
}

export function provideContext<Value, Result>(context: Context<Value>, value: Value, callback: () => Result): Result {
  return createMemo(() => {
    getOwner()!.context = { [context.id]: value };
    return callback();
  })();
}

```

I am wondering if we could refactor out `Provider` from the context, making it a separate component `ContextProvider`, so that the context could be a general mechanism without coupling with view things of solid.

**Comment by ryansolid** [comment]
Is there any blockers just implementing what you have in userspace. The reason I ask is regardless of potentially providing this method I'd still have the component form. This is largely motivated by the content of this article: https://overreacted.io/why-isnt-x-a-hook/. Specificifically why there is no `useProvider`.

If there is nothing stopping you from doing this I'd prefer to leave stuff as is. At least currently. Not to say this might change but the use of UI components is pretty intentional.

**Reply by zhaoyao91** [reply]
Yes, there is nothing preventing implementing them in user space. Just a discussion from the organization/architecture view. solid-js is special from other UI libraries in the way that the reactive primitives is standalone from the view things. They are 2 layers. So I am just curious if `context` could be general and pure enough in the lower layer, while the UI layer is just a comsumer of it.
